third_party/lua-yaml/lyaml.cc

```diff
@@ -819,6 +815,113 @@ static void find_references(struct lua_yaml_dumper *dumper) {
    }
 }
 
+/**
+ * Replace object at `obj_index` on stack with materialized one if it
+ * is found in reftable at `reftable_index` on stack.
+ *
+ * Return true if replace is done and false otherwise.
+ */
+static bool
+materialize_try_reftable(lua_State *L, int obj_index, int reftable_index)
+{
+   assert(reftable_index > 0);
+   if (obj_index < 0)
+      obj_index = lua_gettop(L) + obj_index + 1;
+   lua_pushvalue(L, obj_index);
+   lua_gettable(L, reftable_index);
+   if (!lua_isnil(L, -1)) {
+      /* replace object with materialized one */
+      lua_replace(L, obj_index);
+      return true;
+   }
+   lua_pop(L, 1);
+   return false;
+}
+
+/**
+ * Call luaL_checkfield recursively for object and replace object on stack
+ * with result. Object is at `obj_index` on stack.
+ *
+ * In the process remember the already seen objects in reference table and
+ * check it before materializing next object. Thus the materialized table
+ * keeps the reference structure of the original table. Also the reference
+ * table is checked for luaL_checkfield result so that reference structure
+ * brought by __serialize is kept too. Reference table is at `reftable_index`
+ * on stack.
+ *
+ * Also tables of materialized object keep the hints of __serialize as the
+ * hints affects later serialization.
+ */
+static void
+materialize(struct lua_yaml_dumper *dumper, int obj_index, int reftable_index)
```

Comment:
Just raw idea: wouldn't it be simpler to write the traversal code using Lua (place to `src/box/serializer.lua`) and call it from the C code using `lua_call()`?

I suspect that there may be problems accessing `__serialize` metamethod for cdata, but I didn't look deeper.

Feel free to close this discussion without any actions -- this is just thought around.